R/multivariate_karcher_mean.R

```diff
@@ -1,124 +1,165 @@
+#' each iteration.
+#' - `qun`: A numeric vector storing the cost function of the Karcher mean or
+#' median at each iteration.
+#'
 #' @keywords srvf alignment
+#'
 #' @references Srivastava, A., Klassen, E., Joshi, S., Jermyn, I., (2011). Shape
-#'    analysis of elastic curves in euclidean spaces. Pattern Analysis and Machine
-#'    Intelligence, IEEE Transactions on 33 (7), 1415-1428.
+#'   analysis of elastic curves in euclidean spaces. Pattern Analysis and
+#'   Machine Intelligence, IEEE Transactions on 33 (7), 1415-1428.
+#'
 #' @export
+#'
 #' @examples
 #' out <- multivariate_karcher_mean(beta[, , 1, 1:2], maxit = 2)
 #' # note: use more functions, small for speed
-multivariate_karcher_mean <- function (beta, lambda = 0.0, maxit = 20, ms = "mean")
+multivariate_karcher_mean <- function(beta,
+                                      rotation = FALSE,
+                                      scale = FALSE,
+                                      maxit = 20,
+                                      ms = c("mean", "median"))
 {
-  if(ms!="mean"&ms!="median"){warning("ms must be either \"mean\" or \"median\". ms has been set to \"mean\"",immediate. = T)}
-  if(ms!="median"){ms = "mean"}
+  ms <- rlang::arg_match(ms)
 
-  tmp = dim(beta)
-  n = tmp[1]
-  T1 = tmp[2]
-  time1 <- seq(0,1,length.out=T1)
-  N = tmp[3]
-  q = array(0, c(n, T1, N))
-  for (ii in 1:N) {
-    beta1 = beta[ , , ii]
-    out = curve_to_q(beta1, FALSE)
-    q[, , ii] = out$q
-  }
-
-  mu = q[, , 1]
-  bmu = beta[, , 1]
-  delta = 0.5
-  tolv = 1e-04
-  told = 5 * 0.001
-  itr = 1
-  sumd = rep(0, maxit + 1)
-  sumd[1] = Inf
-  betan = array(0, c(n, T1, N))
-  qn = array(0, c(n, T1, N))
-  normbar = rep(0, maxit + 1)
-  if(ms == "median"){ #run for median only, saves memory if getting mean
-    d_i = rep(0,N) #include vector for norm calculations
-    v_d = array(0, c(n, T1, N)) #include array to hold v_i / d_i
-  }
+  dims <- dim(beta)
+  L <- dims[1] # Dimension of codomain
+  M <- dims[2] # Size of the evaluation grid
+  N <- dims[3] # Sample size
+  q <- array(dim = c(L, M, N)) # Array for hosting SRVFs
 
+  preproc <- furrr::future_map(1:N, \(n) {
+    beta1 <- beta[ , , n]
+    out <- curve_to_q(beta1, scale = scale)
+    q1 <- out$q
+    if (scale)
+      beta1 <- beta1 / out$len
+    list(q1 = q1, beta1 = beta1)
+  }, .options = furrr::furrr_options(seed = TRUE))
 
-  cat("\nInitializing...\n")
-  gam = matrix(0,T1,N)
-  for (k in 1:N) {
-    out = find_rotation_seed_unqiue(mu, q[, , k], 'O', FALSE, FALSE, lambda)
-    gam[, k] = out$gambest
+  for (n in 1:N) {
+    q[ , , n] <- preproc[[n]]$q1
+    beta[ , , n] <- preproc[[n]]$beta1
   }
 
-  gam = t(gam)
-  gamI = SqrtMeanInverse(t(gam))
-  bmu = group_action_by_gamma_coord(bmu, gamI)
-  mu = curve_to_q(bmu, FALSE)$q
-  mu[is.nan(mu)] <- 0
+  # Initialize mean as the pointwise mean of the curves
+  # AST: improve with medoid, meaning compute distance matrix first.
+  qmean <- rowMeans(q, dims = 2)
```

Comment:
Please do not initialize with rowMeans, please pick a curve closest to the pointwise mean.

Reply:
Yeah I figured, hence my comment above starting with AST... I might implement an equivalent of the `stats::dist()` function that uses internally `calc_shape_dist()`. What do you think? If such a function existed, we could initialise with the most central of the curves.

Reply:
That could work and might be a nice feature, also if you look at `time_warping` we initialize the algorithm in this manner as well.